src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+  if (!operator_scan_predicates) {
+    return input_table_statistics;
+  } else {
+    auto output_table_statistics = input_table_statistics;
+
+    for (const auto& operator_scan_predicate : *operator_scan_predicates) {
+      output_table_statistics = estimate_operator_scan_predicate(output_table_statistics, operator_scan_predicate);
+    }
+
+    return output_table_statistics;
+  }
+}
+
+std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_join_node(
+    const JoinNode& join_node, const std::shared_ptr<TableStatistics>& left_input_table_statistics,
+    const std::shared_ptr<TableStatistics>& right_input_table_statistics) {
+  if (join_node.join_mode == JoinMode::Cross) {
+    return estimate_cross_join(*left_input_table_statistics, *right_input_table_statistics);
+  } else {
+    // TODO(anybody) Join cardinality estimation is consciously only performed for the primary join predicate. #
+    const auto primary_operator_join_predicate = OperatorJoinPredicate::from_expression(
+        *join_node.join_predicates()[0], *join_node.left_input(), *join_node.right_input());
+
+    if (primary_operator_join_predicate) {
+      switch (join_node.join_mode) {
+        case JoinMode::Semi:
+        case JoinMode::AntiNullAsTrue:
+        case JoinMode::AntiNullAsFalse:
+          // TODO(anybody) Implement estimation of Semi/Anti joins
+          return left_input_table_statistics;
+
+        // TODO(anybody) For now, handle outer joins just as inner joins
+        case JoinMode::Left:
+        case JoinMode::Right:
+        case JoinMode::FullOuter:
+        case JoinMode::Inner:
+          switch (primary_operator_join_predicate->predicate_condition) {
+            case PredicateCondition::Equals:
+              return estimate_inner_equi_join(primary_operator_join_predicate->column_ids.first,
+                                              primary_operator_join_predicate->column_ids.second,
+                                              *left_input_table_statistics, *right_input_table_statistics);
+
+            // TODO(anybody) Implement estimation for non-equi joins
+            case PredicateCondition::NotEquals:
+            case PredicateCondition::LessThan:
+            case PredicateCondition::LessThanEquals:
+            case PredicateCondition::GreaterThan:
+            case PredicateCondition::GreaterThanEquals:
+            case PredicateCondition::Between:
+            case PredicateCondition::In:
+            case PredicateCondition::NotIn:
+            case PredicateCondition::Like:
+            case PredicateCondition::NotLike:
+              return estimate_cross_join(*left_input_table_statistics, *right_input_table_statistics);
+
+            case PredicateCondition::IsNull:
+            case PredicateCondition::IsNotNull:
+              Fail("IS NULL is an invalid join predicate");
+          }
+          Fail("Unreachable, but GCC doesn't realize...");
```

Comment:
We have a standard text for that. Something like "GCC thinks...". Makes it easier to remove all of these once gcc fixes it.